### PR TITLE
[ADD] Enforce unique product codes on contracts

### PR DESCRIPTION
**Pull request by veloutin**
Not sure if this is too specific, but here goes:

This module adds a restriction on contracts (account.analytic.account) services (contract.service) so that product codes cannot be repeated.
1. Why not a constraint: To allow a smooth transition for existing data, the check is enforced when creating contracts or services, or when modifying contract services. This also prevents the new check from failing invoicing processes or anything that might write to the contract.
2. Why restrict based on product code: We use lines with a "Detail" product that may be repeated for each service/product. This allows enforcing only on actual products.

### CODE REVIEW

File: contract_isp_unique_product/contract.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+from openerp.tools.translate import _
+
+
+def unique(values):
+    seen = set()
+    for v in values:
+        if v in seen:
+            return False
+        seen.add(v)
+    return True
```
Comment (bwrsandman) [inline contract_isp_unique_product/contract.py]: Simpler and more space efficient: 

``` python
def unique(values):
    return not any(i for i in values if values.count(i) > 1)
```

More time efficient O(n) vs O(n²):

``` python
def unique(values):
    return len(values) == len(set(values))
```

**Comment by bwrsandman** [comment]
:+1:

File: contract_isp_unique_product/__openerp__.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+{
+    'name': 'Contract ISP - Unique Product in Contract',
+    'version': '0.1',
+    'author': 'Savoir-faire Linux',
```
Comment (yvaucher) [inline contract_isp_unique_product/__openerp__.py]: Missing OCA in author

**Comment by gurneyalex** [comment]
:+1:

**Pull request merged**